server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTarget.java

```diff
@@ -141,141 +136,115 @@ public void writeFileChunk(
+            source.getSegmentFiles(getId(), checkpointInfo.getCheckpoint(), filesToFetch, indexShard, getFilesListener);
+        }, listener::onFailure);
+
+        getFilesListener.whenComplete(response -> {
+            finalizeReplication(checkpointInfoListener.result());
+            listener.onResponse(null);
+        }, listener::onFailure);
     }
 
-    private void getFiles(CheckpointInfoResponse checkpointInfo, StepListener<GetSegmentFilesResponse> getFilesListener)
-        throws IOException {
+    private List<StoreFileMetadata> getFiles(CheckpointInfoResponse checkpointInfo) throws IOException {
         cancellableThreads.checkForCancel();
         state.setStage(SegmentReplicationState.Stage.FILE_DIFF);
         final Store.RecoveryDiff diff = Store.segmentReplicationDiff(checkpointInfo.getMetadataMap(), indexShard.getSegmentMetadataMap());
-        logger.trace("Replication diff for checkpoint {} {}", checkpointInfo.getCheckpoint(), diff);
+        logger.trace(() -> new ParameterizedMessage("Replication diff for checkpoint {} {}", checkpointInfo.getCheckpoint(), diff));
         /*
          * Segments are immutable. So if the replica has any segments with the same name that differ from the one in the incoming
          * snapshot from source that means the local copy of the segment has been corrupted/changed in some way and we throw an
          * IllegalStateException to fail the shard
          */
         if (diff.different.isEmpty() == false) {
-            IllegalStateException illegalStateException = new IllegalStateException(
+            throw new OpenSearchCorruptionException(
                 new ParameterizedMessage(
                     "Shard {} has local copies of segments that differ from the primary {}",
                     indexShard.shardId(),
                     diff.different
                 ).getFormattedMessage()
             );
-            ReplicationFailedException rfe = new ReplicationFailedException(
-                indexShard.shardId(),
-                "different segment files",
-                illegalStateException
-            );
-            fail(rfe, true);
-            throw rfe;
         }
 
         for (StoreFileMetadata file : diff.missing) {
             state.getIndex().addFileDetail(file.name(), file.length(), false);
         }
-        // always send a req even if not fetching files so the primary can clear the copyState for this shard.
-        state.setStage(SegmentReplicationState.Stage.GET_FILES);
-        cancellableThreads.checkForCancel();
-        source.getSegmentFiles(getId(), checkpointInfo.getCheckpoint(), diff.missing, indexShard, getFilesListener);
+        return diff.missing;
     }
 
-    private void finalizeReplication(CheckpointInfoResponse checkpointInfoResponse, ActionListener<Void> listener) {
+    private void finalizeReplication(CheckpointInfoResponse checkpointInfoResponse) throws OpenSearchCorruptionException {
```

Comment:
Is there any special handling we're adding(or plan to add) for `OpenSearchCorruptionException`? 
We were discussing about recoverable v/s non-recoverable exception during replication earlier so wanted to confirm. 
Exceptions like `CorruptIndexException` would result in  `OpenSearchCorruptionException` which should be non-recoverable. However we're also throwing the `OpenSearchCorruptionException` for generic `Exception` as well(not sure if this would be unrecoverable at blanket level)